release-23.1: roachtest: disable or fix failing azure tests

DarrylWong opened the pull request:
Backport 7/7 commits from #121709.

/cc @cockroachdb/release

---

In the near future we will start running nightly roachtests on Azure. This PR attempts to disable or fix most roachtests that fail on Azure.

The vast majority of failures are due to some sort of cloud incompatibility, i.e. trying to write to a gcs bucket or using google pub/sub changefeeds. In these cases, the test was disabled for Azure. A few tests however, failed due to a lack of setup and were fixed instead.

We also make a change to when the local cluster cache can delete clusters. Instead, we now attempt to clear the cache of deleted clusters at the start of a roachtest run, but do not allow deletion after that. This is due to Azure returning stale information about which VMs we have running. With the original cache behavior, this would see us removing all Azure VMs after `azure.List` returned no VMs.

Release note: none
Fixes: none
Epic: none

Release Justification: test-only change

Commit:
roachtest: avoid nil logger

When creating a new cluster, this moves the initialisation
of the log with retry number to the top of the loop,
so that we can pass in the log reference to the `clusterImpl`.

Without this, new clusters are susceptible to a nil pointer.

This surfaced when testing Azure cloud, since not all provider
functions are implemented, and a log statement is issued.

Epic: none
Release note: none

DarrylWong committed:
roachtest: add azure configs for tpccbench and tpcc/headroom tests

These tests depend on a per cloud config to determine
the number of warehouses to load in. The numbers were
based off of gce but in the future we should tune the
number after we have more data.

DarrylWong committed:
roachtest: disable certain cdc sink tests for azure

Several CDC tests rely on streaming to a pubsub or
cloudstorage sink. On azure, we do not have the permissions
currently to do this so we disable them on azure.

DarrylWong committed:
roachtest: disable backup tests on azure

Most backup tests are currently not set up for azure.

DarrylWong committed:
roachtest: disable disk stall tests for azure

Some roachtests simulate disk stalls by using dmsetup
or cgroupsv2. Configs need to be added to Azure for both.
However, the former runs into "Device or resource busy"
and the latter does not properly stall the disk, so
both are disabled for azure until fixed.

DarrylWong committed:
roachtest: fix ORM tests for azure

Some tests are not compatible with azure due
to requiring port service registration.

DarrylWong committed:
roachtest: disable tests that use fixture buckets on azure

These tests require a fixture bucket which is only configured
on gce right now.

DarrylWong committed:
roachtest: add clear-cluster-cache flag

This flag indicates if we should remove missing clusters
from the local cache when syncing resources with providers.

Because Azure can sometimes return stale VM information
when many PUT calls are made in succession, we cannot
trust that missing clusters in ListCloud are actually
deleted.

This flag allows azure runs that spin up multiple
concurrent clusters, i.e. azure nightlies, to
disable clearing the cache.

DarrylWong committed:
roachtest: add missing backslash in nightly script

This causes all subsequent flags to be dropped.
Fixes: #123348
Release note: none
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123670)
<!-- Reviewable:end -->

DarrylWong committed:
roachtest: disable azure tests that use zfs or specify ssd counts

Both these options are not supported for roachprod azure yet,
resulting in cluster_creation errors.

Informs: #120621
Fixes: none
Epic: none
Release note: none

yuzefovich commented:
courtesy cleanup
